File: apps/scraper/app/sp_api/catalog.py
"""
SP-API CatalogItems 2022-04-01
==============================
Retrieves product identity, images, category rankings and product-type for
a single ASIN via ``GET /catalog/2022-04-01/items/{asin}``.

Docs: https://developer-docs.amazon.com/sp-api/reference/get-catalog-item
"""
from __future__ import annotations

from app.sp_api.client import MARKETPLACE_IDS, is_configured, sp_request

# Data groups we want back from the API
_INCLUDED_DATA = "attributes,images,productTypes,summaries,salesRanks"


# ---------------------------------------------------------------------------
# Public interface
# ---------------------------------------------------------------------------

async def get_catalog_item(asin: str, marketplace: str = "US") -> dict:
    """
    Return normalised catalog data for *asin* in *marketplace*.

    When SP-API credentials are absent, returns a minimal stub dict so the
    combined lookup can still be assembled.
    """
    if not is_configured():
        return _stub_catalog(asin, marketplace)

    marketplace_id = MARKETPLACE_IDS.get(marketplace, "ATVPDKIKX0DER")

    data = await sp_request(
        "GET",
        marketplace,
        f"/catalog/2022-04-01/items/{asin}",
        params={
            "marketplaceIds": marketplace_id,
            "includedData": _INCLUDED_DATA,
        },
    )

    return _parse(data, asin, marketplace, marketplace_id)


# ---------------------------------------------------------------------------
# Parsing
# ---------------------------------------------------------------------------

def _parse(data: dict, asin: str, marketplace: str, marketplace_id: str) -> dict:
    # ── Summaries ──────────────────────────────────────────────────────────
    summaries: list[dict] = data.get("summaries", [])
    summary = (
        next((s for s in summaries if s.get("marketplaceId") == marketplace_id), None)
        or (summaries[0] if summaries else {})
    )

    # ── Main image ─────────────────────────────────────────────────────────
    images_sets: list[dict] = data.get("images", [])
    img_list: list[dict] = next(
        (g.get("images", []) for g in images_sets if g.get("marketplaceId") == marketplace_id),
        [],
    )
    main_image: str | None = next(
        (img["link"] for img in img_list if img.get("variant") == "MAIN"),
        None,
    )

    # ── Sales ranks ────────────────────────────────────────────────────────
    ranks_sets: list[dict] = data.get("salesRanks", [])
    class_ranks: list[dict] = next(
        (r.get("classificationRanks", []) for r in ranks_sets if r.get("marketplaceId") == marketplace_id),
        [],
    )
    display_ranks: list[dict] = next(
        (r.get("displayGroupRanks", []) for r in ranks_sets if r.get("marketplaceId") == marketplace_id),
        [],
    )
    bsr_entry = class_ranks[0] if class_ranks else (display_ranks[0] if display_ranks else {})

    # ── Product type ───────────────────────────────────────────────────────
    product_types: list[dict] = data.get("productTypes", [])
    product_type: str | None = product_types[0].get("productType") if product_types else None

    return {
        "asin": asin,
        "title": summary.get("itemName"),
        "brand": summary.get("brand"),
        "manufacturer": summary.get("manufacturer"),
        "model_number": summary.get("modelNumber"),
        "product_type": product_type,
        "main_image": main_image,
        "bsr": bsr_entry.get("rank"),
        "bsr_category": bsr_entry.get("title") or bsr_entry.get("displayName"),
        "marketplace": marketplace,
        "source": "sp-api",
    }


# ---------------------------------------------------------------------------
# Stub
# ---------------------------------------------------------------------------

def _stub_catalog(asin: str, marketplace: str) -> dict:
    return {
        "asin": asin,
        "title": None,
        "brand": None,
        "manufacturer": None,
        "model_number": None,
        "product_type": None,
        "main_image": None,
        "bsr": None,
        "bsr_category": None,
        "marketplace": marketplace,
        "source": "stub",
    }

File: apps/scraper/sp-api/catalog.ts
import { CONFIG } from '../config'
import { spApiCall, sleep, parseDimensions } from './client'
import type { CatalogItemResult, SpApiDimensions } from './types'

// ─── RAW CATALOG API RESPONSE ────────────────────────────────────────────────

interface RawCatalogResponse {
    asin: string
    summaries?: Array<{
        marketplaceId: string
        itemName?: string
        brand?: string
        manufacturer?: string
        modelNumber?: string
        packageQuantity?: number
        color?: string
        browseClassification?: { classificationId: string; displayName: string }
        websiteDisplayGroup?: string
        listingDate?: string  // ISO date string e.g. '2023-11-14'
    }>
    attributes?: Record<string, Array<{ value: unknown; language_tag?: string; marketplace_id?: string }>>
    dimensions?: Array<{
        marketplaceId: string
        item?: SpApiDimensions
        package?: SpApiDimensions
    }>
    images?: Array<{
        marketplaceId: string
        images: Array<{ variant: string; link: string; height: number; width: number }>
    }>
    productTypes?: Array<{ marketplaceId: string; productType: string }>
    salesRanks?: Array<{
        marketplaceId: string
        classificationRanks?: Array<{ classificationId: string; title: string; rank: number; link?: string }>
        displayGroupRanks?: Array<{ websiteDisplayGroup: string; title: string; rank: number; link?: string }>
    }>
}

// ─── getCatalogItem ───────────────────────────────────────────────────────────

export async function getCatalogItem(asin: string): Promise<CatalogItemResult | null> {
    const includedData = [
        'summaries',
        'attributes',
        'dimensions',
        'images',
        'productTypes',
        'salesRanks',
    ].join(',')

    const url =
        `https://sellingpartnerapi-na.amazon.com/catalog/2022-04-01/items/${asin}` +
        `?marketplaceIds=${CONFIG.sp_marketplace_id}&includedData=${includedData}&locale=en_US`

    await sleep(CONFIG.spapi_delay_ms)

    const raw = await spApiCall<RawCatalogResponse>(url)
    if (!raw) return null

    // ── Summaries ──────────────────────────────────────────────────────────────
    const summary =
        raw.summaries?.find(s => s.marketplaceId === CONFIG.sp_marketplace_id) ?? raw.summaries?.[0]

    // ── Attributes ─────────────────────────────────────────────────────────────
    const attrs = raw.attributes ?? {}

    const bullet_points: string[] = (attrs['bullet_point'] ?? [])
        .filter(b => !b.marketplace_id || b.marketplace_id === CONFIG.sp_marketplace_id)
        .map(b => String(b.value ?? '').trim())
        .filter(Boolean)

    // Helper: pick attribute value for current marketplace (or first available)
    const attrVal = (key: string): unknown | null => {
        const arr = attrs[key] ?? []
        const byMkt = arr.find(a => a.marketplace_id === CONFIG.sp_marketplace_id)
        const any = arr.find(a => a.value != null)
        return (byMkt ?? any)?.value ?? null
    }

    // list_price from attributes (most reliable price source)
    const rawListPrice = attrVal('list_price') as { value?: number; currency?: string } | number | null
    const list_price =
        rawListPrice != null
            ? typeof rawListPrice === 'object' && rawListPrice !== null
                ? typeof rawListPrice.value === 'number' ? Math.round(rawListPrice.value * 100) / 100 : null
                : typeof rawListPrice === 'number' ? Math.round(rawListPrice * 100) / 100 : null
            : null

    // ── Dimensions ─────────────────────────────────────────────────────────────
    const dimsEntry =
        raw.dimensions?.find(d => d.marketplaceId === CONFIG.sp_marketplace_id) ?? raw.dimensions?.[0]
    const itemDims = parseDimensions(dimsEntry?.item)
    const pkgDims = parseDimensions(dimsEntry?.package)

    // ── Images ─────────────────────────────────────────────────────────────────
    const imagesEntry =
        raw.images?.find(i => i.marketplaceId === CONFIG.sp_marketplace_id) ?? raw.images?.[0]
    // Prefer largest available — sort by pixel area descending
    const sortedImages = imagesEntry?.images
        ?.filter(i => i.variant === 'MAIN')
        .sort((a, b) => b.height * b.width - a.height * a.width)
    const main_image_url = sortedImages?.[0]?.link ?? null

    // ── Product types ──────────────────────────────────────────────────────────
    const pt =
        raw.productTypes?.find(p => p.marketplaceId === CONFIG.sp_marketplace_id) ?? raw.productTypes?.[0]

    // ── Sales ranks ────────────────────────────────────────────────────────────
    const ranksEntry =
        raw.salesRanks?.find(r => r.marketplaceId === CONFIG.sp_marketplace_id) ?? raw.salesRanks?.[0]
    // category_ranks: classification ranks only (real browse node IDs that exist in
    // amazon_categories). displayGroupRanks use string keys like
    // 'fashion_display_on_website' that are not in our DB.
    const category_ranks = (ranksEntry?.classificationRanks ?? []).map(r => ({
        classificationId: r.classificationId,
        rank: r.rank,
    }))

    return {
        title: summary?.itemName ?? null,
        brand: summary?.brand ?? null,
        manufacturer: summary?.manufacturer ?? null,
        model_number: summary?.modelNumber ?? null,
        package_quantity: summary?.packageQuantity ?? null,
        color: summary?.color ?? null,
        list_price,
        main_image_url,
        bullet_points,
        product_type: pt?.productType ?? null,
        browse_node_id: summary?.browseClassification?.classificationId ?? null,

        item_length_cm: itemDims.length_cm,
        item_width_cm: itemDims.width_cm,
        item_height_cm: itemDims.height_cm,
        item_weight_kg: itemDims.weight_kg,
        pkg_length_cm: pkgDims.length_cm,
        pkg_width_cm: pkgDims.width_cm,
        pkg_height_cm: pkgDims.height_cm,
        pkg_weight_kg: pkgDims.weight_kg,

        category_ranks,
        listing_date: (() => {
            const raw = (attrs['product_site_launch_date'] ?? [])[0]?.value
            if (!raw) return null
            const d = new Date(raw as string)
            return isNaN(d.getTime()) ? null : d.toISOString().slice(0, 10)
        })(),
    }
}
